### PR TITLE
Materialized views: fix possibly old views coming from other nodes

### PR DESCRIPTION
**Pull request by eliransin**
Migration manager has a function to get a schema (for read or write),
this function queries a peer node and retrieves the schema from it. One
scenario where it can happen is if an old node, queries an old not fixed
index.
This makes a hole through which views that are only adjusted for reading
can slip through.

Here we plug the hole by fixing such views before they are registered.

### CODE REVIEW

**Comment by eliransin** [comment]
@psarna / @nyh can you give this a priority? It is a blocker for one of our enterprise releases.

**Comment by psarna** [comment]
ok, reviewing now

File: service/migration_manager.cc [hunk 1]
```diff
@@ -1090,6 +1091,7 @@ future<schema_ptr> get_schema_definition(table_schema_version v, netw::messaging
                 // It should be fine since if we tried to register a view for which
                 // we don't know the base table, our registry is broken.
                 schema_ptr base_schema = db.find_schema(s->view_info()->base_id());
+                s = db::schema_tables::maybe_fix_legacy_secondary_index_mv_schema(get_local_storage_proxy().local_db(), view_ptr(s), base_schema, db::schema_tables::preserve_version::yes);
```
Comment (psarna) [inline service/migration_manager.cc:1094]: I think you can't just unconditionally take the return value from here, you need to check if it's not null. Here's an excerpt from the implementation:
https://github.com/scylladb/scylla/blob/dbd0b9a3efe8eaaa436586bfa9be050a601915ad/db/schema_tables.cc#L3119-L3122
So in the (quite likely I think) case when the schema is ok, the line below would dereference a null pointer.

Comment (eliransin) [reply service/migration_manager.cc:1094]: Ohh you are right. Will fix it right away.

File: service/migration_manager.cc [hunk 2]
```diff
@@ -1090,6 +1091,10 @@ future<schema_ptr> get_schema_definition(table_schema_version v, netw::messaging
                 // It should be fine since if we tried to register a view for which
                 // we don't know the base table, our registry is broken.
                 schema_ptr base_schema = db.find_schema(s->view_info()->base_id());
+                auto fixed_view = db::schema_tables::maybe_fix_legacy_secondary_index_mv_schema(get_local_storage_proxy().local_db(), view_ptr(s), base_schema, db::schema_tables::preserve_version::yes);
+                if (fixed_view) {
+                    s = fixed_view;
```
Comment (psarna) [inline service/migration_manager.cc:1096]: I looked at the callers of `get_schema_definition()` and I don't really understand if and how this code will work. Schema pointers which describe the existing tables in a Scylla node have a special `_registry_entry` field which points to the schema registry entry. This entry (as far as I know) is only allocated for schema instances which were loaded from disk or announced via the migration manager, and is a `nullptr` otherwise. Here, we just build a temporary `schema_ptr` instance (`fixed_view`) and return it as if it was one of the registered schemas. The potential problem I see is that e.g. here:
https://github.com/scylladb/scylla/blob/dc7beec382535d9c55b0b6523ab5d2b9c94686d7/service/migration_manager.cc#L338-L340
we take the return value from `get_schema_definition` and dereference its registry entry, assuming that it exists. But does it, if we just generated this schema ourselves without registering it anywhere?

By the way - how was this change tested, and is there an issue I can read to get more context on why it's needed?

/cc @tgrabiec

Comment (eliransin) [reply service/migration_manager.cc:1096]: @psarna you are right, it smells bad. I will look into it.

Comment (eliransin) [reply service/migration_manager.cc:1096]: You are right, it is a problem, I should fix the view prior to registering it. I will make the changes.

File: service/migration_manager.cc [hunk 3]
```diff
@@ -1076,16 +1077,27 @@ future<schema_ptr> get_schema_definition(table_schema_version v, netw::messaging
...
+                return frozen_schema{us};
             });
         });
     }).then([] (schema_ptr s) {
         // If this is a view so this schema also needs a reference to the base
         // table.
         if (s->is_view()) {
             if (!s->view_info()->base_info()) {
-                auto& db = service::get_local_storage_proxy().get_db().local();
+                auto& db = service::get_local_storage_proxy().local_db();
```
Comment (psarna) [inline service/migration_manager.cc:1100]: nitpick: unrelated change, might be a potential minor merging issue if it creates conflicts

**Review by psarna — APPROVED** [review]
Looks like the proper place, but I'd really prefer somebody more experience with schema registry to ack as well, since I don't exactly grasp what happens in the original code either.

File: service/migration_manager.cc (hunk 3 above)
Comment (eliransin) [reply service/migration_manager.cc:1100]: will change it back.

**Comment by eliransin** [comment]
> Looks like the proper place, but I'd really prefer somebody more experience with schema registry to ack as well, since I don't exactly grasp what happens in the original code either.

@tgrabiec can you please assist?

**Comment by psarna** [comment]
I also leave this pull request as "approved" by me, so whoever acks later, feel free to merge

**Comment by eliransin** [comment]
@tgrabiec ping. This is a blocker for enterprise.

File: service/migration_manager.cc (hunk 3 above)
Comment (nyh) [reply service/migration_manager.cc:1100]: I think you didn't change it back? (at least I don't see where)?

**Review by nyh — APPROVED** [review]

**Comment by avikivity** [comment]
Are there vulnerable branches? Is so file an issue so I can backport.

**Comment by eliransin** [comment]
@avikivity opened #8554, 4.3 and later are vulnerable. I will label the issue.

Just for completeness I will be more accurate here:
If upgrading from a version with the fix for example 4.3 after backport to 4.4, this 4.4 version even without the fix is 
not vulnerable because the rogue views will be fixed during the upgrade to 4.3 + fix.

**Comment by avikivity** [comment]
I see. Still, it's best to stick with the policy that there are no gaps in backports (we may choose not to backport to an old version, but if we backport to N, all higher versions are also patched).
